README.md
# RBX-DQ-AutoLobby

Auto-Lobby Script for the game "Dungeon Quest" on Roblox

## Introduction

   This script was designed to give you the additional capability of **Auto-hosting** a game lobby, or **Auto-joining** another player's game lobby. It was made to work with Blake's Autofarm on the game Dungeon Quest. It is useful for auto-hosting/joining friends and alt accounts.

## Usage

   To use this script you must **download `AutolobbyDQ.lua`** and then **edit the settings** in the config section at the top of the file. What each setting means and how to use it is explained below in Settings section.

   ~~There's two ways you can configure the script to work with the Autofarm script.~~ Autoexec method has been removed due to unreliability.

   **Loadfile Method**: After configuring the file, you **place it in the autoexec folder of your exploit program** (*like Synapse X*), **and place Blake's Autofarm file in the exploit 'workspace' folder**. They should be individual/separate files. You can name the autofarm whatever you wish, just be sure to update the exact name of it in the Misc area of the autolobby file's settings.

   The settings in the Autolobby script do not affect the settings of the Autofarm script nor do they change how it performs. The Autofarm script will detect where it is once a dungeon or boss raid dungeon has loaded, and will do what it is supposed to do from there. *This script only handles lobby creation, whitelisting, and joining.*

   The Autolobby script is now also capable of handling multiple accounts at once on the same PC, using the Multiple_Roblox exploit. Using this tactic, you can host and carry the accounts of your choosing simultaneously. This means on one PC you can have a host auto-creating lobbies as well as one or more alt/carry accounts auto-joining it. For more info read [Multiple_Roblox-Usage.md](https://github.com/dns2/RBX-DQ-AutoLobby/blob/master/Multiple_Roblox-Usage.md)

## Settings

Important: **Usernames**, **Map name**, & **Difficulty** are all **case-sensitive**. *Example: "player1" is not the same as "Player1")

### Script

* Enabled

  `true` : *Enables Autolobby script. Script will attempt to run with configured settings.*

  `false` : *Disables Autolobby. Script will exit immediately.*

* VIPONLY

  **Note**: *This script cannot teleport you to VIP servers. This setting is only for usage with the AutoRestart program. It simply detects whether you have loaded into a VIP server or not and then pauses further script execution based on whether you have enabled it or not.*

  `true` : *Will pause any further execution of script detects you are in a public server. Will prevent the script from continuing.*

  `false` : *Will not suspend the script if it detects you're in a public server when loaded.*

___

### Hosting

* LobbyHost

   The username of the player that will be creating and hosting the game lobby. If you will be the one hosting, then you can either put your own username or simply leave it empty (such as `""`). *If you will be using the Multiple_Roblox method then the exact username is required.*

  **Examples**:

   `"MyUserName"` : *Your username or another player you want to automatically join*

   `""` : *A shortcut for you to host without inputting your username in the script*

* WLParty

   The username(s) to whitelist. Whatever players you put here will be whitelisted for the game you create. This option does nothing if you are not the host. When you are the host and have players listed here, the script will not start the game until all players have been whitelisted and joined the game party.

  **Examples**:

  `{"Player1","Player2","Player3"}` : *Whitelist multiple players*

  `{"Player1"}` : *Whitelist single player*

  `{}` : *Do not whitelist or wait for other players*

* LobbyType

   The type of lobby to create. This option is only used if you will be the one hosting.

  **Examples**:

  `"Dungeon"` : *Create a dungeon lobby*

  `"BossRaid"` : *Create a boss raid lobby*

___

### Dungeon

* MapName

   The name of the Dungeon map. *This option only matters if you are hosting.*

  **Example**:

  `"Desert Temple"` : *To host a dungeon on the Desert Temple map*

* Difficulty

   The difficulty setting for the Dungeon. *This option only matters if you are hosting.*

  **Example**:

  `"Nightmare"` : *To set difficulty to Nightmare*

* Hardcore

   Whether the hardcore setting will be enabled or not for the Dungeon map. *This option only matters if you are hosting.*

  **Example**:

  `true` or `false` : *Enable or disable Hardcore mode*

* WaveDefense

   This option should be kept `false`. (Wave defence doesn't currently work on Blake's Autofarm).

___

### BossRaid

* BRTier

   The tier number of the boss raid lobby you wish to create. *This option only matters if you are hosting.*

  **Example**:

  `10` : *To set the Boss Raid Tier to level 10*

* BRTierReq

   The tier requirement for players joining your boss raid lobby. *This option should remain at `0`. Reserved for a possible future update.*

___

### Misc

* Private

   **(Unused)** Whether or not to make your lobby private or public when creating it. *Keep this set to `false`. Reserved for a possible future.* update.

* DisableMusic

   If `true` this will disable the game music that plays once you have spawned into the game. If `false`, does nothing.

* DisableTrade

   If `true` this will disable your trade requests after spawning into the game. If `false`, leaves trading enabled.

* DebugOutput

   `true` or `false` : Enables or Disables the script output to Roblox Developer Console. *On PC you can quickly open the console using the F9 button.*

* NameOfAutofarmFile

   `"AutofarmFileName.lua"` : The **exact** filename you have for Blake's Autofarm file. *The file must be located in your exploit workspace folder*.

* RunAfOnlyAsHost

   `true` Autofarm file will only be run after the dungeon is loaded **if it's the account hosting the game lobby**

   `false` Autofarm file **will always be allowed to run after a dungeon is loaded**, even if you are just auto-joining another account



CHANGELOG.md
## Changelog

**6/23/20**
* Removed `autoexec` method; `loadfile` is now the sole/primary method.
* Added capability to use with "Multi_Roblox" exploit. [Multiple_Roblox-Usage.md](https://github.com/dns2/RBX-DQ-AutoLobby/blob/master/Multiple_Roblox-Usage.md)
* Added `RunAfOnlyAsHost` option. (Details in README)


DQAutoLobby-dns.lua
-----------------------------------------------------------------------------
-- [[ Dungeon Quest AutoLobby Addon v1.1 - by dns (dns#3420 on discord) ]] --
-----------------------------------------------------------------------------
-- This adds Carrying (Auto-hosting/Auto-joining) functionality to Blake's Autofarm.
-- TO USE: Paste Blake's Autofarm code below this one (or paste this one above Blake's)
-- SUPPORT: If you need help with using/configuring this, then message me (dns#3420) on discord.
-- NOTE: This will not teleport you to VIP servers. The VIPONLY option is only for usage with the AutoRestart program.

-- [[ CONFIG ]] -- (NOTES: Usernames, Map name, & Difficulty are all case-sensitive. *Example: "player1" is not the same as "Player1")
---> Hosting options
_G.DQAL_LobbyHost = "HostName" -- Put your username here if you'll be hosting/carrying other players, or the host's name if you'll be joining someone else's game
_G.DQAL_WLParty = {"Player1"} -- Username(s) to whitelist. Example: {"Player1","Player2","Player3"} | Or for solo farming just use {}
_G.DQAL_LobbyType = "Dungeon" -- "Dungeon" or "BossRaid"
---> Dungeon options
_G.DQAL_MapName = "Orbital Outpost"
_G.DQAL_Difficulty = "Nightmare"
_G.DQAL_Hardcore = true
_G.DQAL_WaveDefense = false -- Wave defence doesn't currently work on Blake's Autofarm. Leave this false, for now.
---> BossRaid options
_G.DQAL_BRTier = 30 -- I'll add an Auto feature for this in a future update.
_G.DQAL_BRTierReq = 0 -- Leave at 0. This option is for a future update.
---> Universal options
_G.DQAL_Enabled = true -- Change to false to disable/skip the Autolobby code. (Will make it skip this and go straight to Blake's Autofarm)
_G.DQAL_Private = true -- Keep this true, will be used in a future update. (False allows open lobbies which anyone can join)
_G.DQAL_VIPONLY = false -- true = Suspend everything unless on a VIP Server (Only used for autofarming with autorestart program) | false = Continue working on public servers (risky)
_G.DQAL_DisableMusic = true -- true = Music disabled | false = Music enabled
_G.DQAL_DisableTrade = true -- true = Trading disabled | false = Trading enabled
_G.DQAL_VerboseDebug = true -- Enable or Disable information/error output to Roblox Developer Console

function Prnt(_msg)
	if _G.DQAL_VerboseDebug then print("[DQAL]: "..tostring(_msg)) end
end
Prnt("Dungeon Quest AutoLobby has loaded, Waiting for game to load..")
repeat wait() until game:IsLoaded()
Prnt("Game is loaded.")
local MyName = game.Players.LocalPlayer.Name
local Proceed = false
-- [[ FUNCTIONS ]]
function findInTable(tbl, thing)
    for _, x in ipairs(tbl) do
        if x == thing then
            return true
		end
    end
	return false
end
function whitelist()
	if #_G.DQAL_WLParty <= 0 then return false end
	local myLobby = {}
	local plyrsAdded = {}
	local plyrsOnline = {}
	local partyReady = false
	Prnt("Waiting to whitelist ["..#_G.DQAL_WLParty.."] players..")
	repeat
		wait()
		if _G.DQAL_LobbyType == "Dungeon" then
			myLobby = game.Workspace.games.inLobby[MyName]:GetChildren()
			if #_G.DQAL_WLParty == #myLobby-2 then partyReady = true end
		elseif _G.DQAL_LobbyType == "BossRaid" then
			myLobby = game.Workspace.bossLobbies[MyName].players:GetChildren()
			if #_G.DQAL_WLParty == #myLobby-1 then partyReady = true end
		end
		plyrsOnline = game.Players:GetChildren()
		for i, v in ipairs(plyrsOnline) do
			if findInTable(_G.DQAL_WLParty, v.Name) and not findInTable(myLobby, v.Name) and not findInTable(plyrsAdded, v.Name) then
				if _G.DQAL_LobbyType == "Dungeon" then
					game.ReplicatedStorage.remotes.addPlayerToWhitelist:FireServer(v.Name)
				elseif _G.DQAL_LobbyType == "BossRaid" then
					game.ReplicatedStorage.remotes.addPlayerToBossWhitelist:FireServer(v.Name)
				end
				table.insert(plyrsAdded, v.Name)
				Prnt(v.Name.." whitelisted. ["..#plyrsAdded.."/"..#_G.DQAL_WLParty.."]")
			end
		end
	until(partyReady and (#_G.DQAL_WLParty == #plyrsAdded))
	return true
end

if _G.DQAL_Enabled and game.PlaceId == 2414851778 then
	game.ReplicatedStorage.remotes.loadPlayerCharacter:FireServer(true)
	Prnt("Waiting for character to load..")
	game.Players.LocalPlayer.CharacterAdded:Wait()
	Prnt("Character loaded.")
	MyName = game.Players.LocalPlayer.Name
	Prnt("Hello, "..MyName..". Dungeon Quest Lobby found, continuing..")

	--disable music/trading
	if _G.DQAL_DisableMusic then game.ReplicatedStorage.remotes.changeMute:FireServer() end
	if _G.DQAL_DisableTrade then game.ReplicatedStorage.remotes.changeDnD:FireServer() end

	if _G.DQAL_VIPONLY and not game.Workspace.vipServer.Value then
		Prnt("VIPONLY is enabled, and this is a Public server. Stopping here.")
		repeat wait() until not _G.DQAL_VIPONLY
	end

	if _G.DQAL_LobbyHost == MyName then
		--create the lobby
		if _G.DQAL_LobbyType == "Dungeon" then
			Prnt("Creating your Dungeon Lobby..")
			game.ReplicatedStorage.remotes.createLobby:InvokeServer(_G.DQAL_MapName, _G.DQAL_Difficulty, 0, _G.DQAL_Hardcore, _G.DQAL_Private, _G.DQAL_WaveDefense)
		elseif _G.DQAL_LobbyType == "BossRaid" then
			Prnt("Creating your Boss Raid Lobby..")
			game.ReplicatedStorage.remotes.createBossLobby:InvokeServer(_G.DQAL_BRTier, _G.DQAL_Private, _G.DQAL_BRTierReq)
		end

		--whitelist players
		whitelist()

		--start the dungeon
		Prnt("All players have joined. Starting game..")
		if _G.DQAL_LobbyType == "Dungeon" then
			game.ReplicatedStorage.remotes.startDungeon:FireServer()
		elseif _G.DQAL_LobbyType == "BossRaid" then
			game.ReplicatedStorage.remotes.startBossRaid:FireServer()
		end

		--wait for teleport
		Prnt("Waiting to teleport out..")
		while game.PlaceId == 2414851778 do
			wait()
		end

	else -- if joining another host

		--get lobby
		local lobbies = {}
		Proceed = false
		Prnt("Searching for ".._G.DQAL_LobbyHost.."'s Lobby..")
		repeat
			wait()
			if _G.DQAL_LobbyType == "Dungeon" then
				lobbies = game.Workspace.games.inLobby:GetChildren()
			elseif _G.DQAL_LobbyType == "BossRaid" then
				lobbies = game.Workspace.bossLobbies:GetChildren()
			end
			for i,v in next, lobbies do
				if tostring(v) == _G.DQAL_LobbyHost then
					Proceed = true
					Prnt("Lobby found.")
					break
				end
			end
		until(Proceed)

		--join lobby
		Prnt("Joining ".._G.DQAL_LobbyHost.."'s Lobby..")
		Proceed = false
		if _G.DQAL_LobbyType == "Dungeon" then
			game.ReplicatedStorage.remotes.joinDungeon:InvokeServer(_G.DQAL_LobbyHost)
		elseif _G.DQAL_LobbyType == "BossRaid" then
			local raidHost = game.Workspace.bossLobbies[_G.DQAL_LobbyHost]
			game.ReplicatedStorage.remotes.playerJoinBossLobby:InvokeServer(raidHost)
		end
		repeat
			wait()
			local inlobby = {}
			if _G.DQAL_LobbyType == "Dungeon" then
				inlobby = game.Workspace.games.inLobby[_G.DQAL_LobbyHost]:GetChildren()
			elseif _G.DQAL_LobbyType == "BossRaid" then
				inlobby = game.Workspace.bossLobbies[_G.DQAL_LobbyHost].players:GetChildren()
			end
			for i,v in next, inlobby do
				if tostring(v) == game.Players.LocalPlayer.Name then
					Proceed = true
					Prnt("Lobby joined.")
					break
				end
			end
		until(Proceed)

		--wait for teleport
		Prnt("Waiting to teleport out..")
		while game.PlaceId == 2414851778 do
			wait()
		end
	end
else
	Prnt("Not a Dungeon Quest Lobby, or Autolobby is disabled. Exiting.")
end

-- // End of Dungeon Quest AutoLobby. Paste Blake's Autofarm code (ALL of it, including Settings) below this line // --


AutolobbyDQ[dns].lua
-----------------------------------------------------------------------------
-- [[ Dungeon Quest AutoLobby Addon v1.2 - by dns (dns#3420 on discord) ]] --
--               https://github.com/dns2/RBX-DQ-AutoLobby                  --
-----------------------------------------------------------------------------
--[[ 	FOR: Adds Carrying (Auto-hosting or Auto-joining) functionality to Blake's Autofarm.
		TO USE: Configure the settings below and place in autoexec folder along with Blake's Autofarm (separately).
		SUPPORT: If you need help with using/configuring this, then message me (dns#3420) on discord, or use link below.
		* Full Configuration and Usage info is available at https://github.com/dns2/RBX-DQ-AutoLobby/blob/master/README.md
		NOTES:
			This will not teleport you to VIP servers. The VIPONLY option is only for usage with the 'AutoRestart' program.
			Usernames, Map name, & Difficulty are all case-sensitive. *Example: "player1" is not the same as "Player1")		]]

-- [ CONFIG ]
Settings = {
--> [ Script ]
	Enabled = true,						-- false = disable/skip Autolobby
	VIPONLY = false,					-- true = Suspend everything unless on a VIP Server
--> [ Hosting ]
	LobbyHost = "HostName",				-- Username of player that will be creating/hosting the game lobby. Will also accept "" to be host
	WLParty  = {"Player1"},				-- Username(s) to whitelist. Example: {"Player1","Player2","Player3"} | Will also accept {} for solo
	LobbyType = "BossRaid",				-- "Dungeon" or "BossRaid"
--> [ Dungeon ]
	MapName = "Orbital Outpost",
	Difficulty = "Nightmare",
	Hardcore = true,
	WaveDefense = false,				-- Keep this false. (Wave defence doesn't currently work on Blake's Autofarm)
--> [ BossRaid ]
	BRTier = 30,						-- I'll add an Auto feature for this in a future update
	BRTierReq = 0,						-- Leave at 0. Reserved for future update
--> [ Misc ]
	Private = true,						-- Keep this true. Reserved for future update
	DisableMusic = true,				-- true = Music disabled | false = Music enabled
	DisableTrade = true,				-- true = Trading disabled | false = Trading enabled
	DebugOutput = true					-- Enable or Disable script output to Roblox Developer Console
} -- [ END CONFIG ] --> (Do not edit below this line)

DQAL = {}
SRemote = {
	rFile = nil,
	invoke = {"createLobby","createBossLobby","joinDungeon","playerJoinBossLobby"},
	fire = {"loadPlayerCharacter","changeMute","changeDnD","addPlayerToWhitelist","addPlayerToBossWhitelist","startDungeon","startBossRaid"}	}
setmetatable(SRemote, {__call =
function(tbl, rScript, ...)
	if not tbl.rFile then tbl.rFile = game:GetService("ReplicatedStorage"):FindFirstChild("remotes") end
	if table.find(tbl.invoke, rScript) then tbl.rFile:FindFirstChild(rScript):InvokeServer(...) return end
	if table.find(tbl.fire, rScript) then tbl.rFile:FindFirstChild(rScript):FireServer(...) return end
end})

function DQAL.Prnt(...)
	if Settings.DebugOutput then print("[DQAL]:",...) end
end

function DQAL.Die(p, ...)
	if p then DQAL.Prnt(...) end
	while script do script:Destroy() end
end

function DQAL.PreRun()
	DQAL.Prnt("Dungeon Quest AutoLobby initialized. Waiting for game..")
	repeat wait() until game:IsLoaded()
	if game.PlaceId ~= 2414851778 then DQAL.Die(true, "Not a Dungeon Quest Lobby. Exiting..") end
	DQAL.Prnt("Game is loaded. Waiting for character..")
	SRemote("loadPlayerCharacter", true)
	pChar = game.Players.LocalPlayer.Character or game.Players.LocalPlayer.CharacterAdded:wait()
	DQAL.Prnt("Character loaded. Continuing..")
	MyName = game.Players.LocalPlayer.Name
	if Settings.LobbyHost == "" then Settings.LobbyHost = MyName end
	if Settings.DisableMusic then SRemote("changeMute") end
	if Settings.DisableTrade then SRemote("changeDnD") end
	if Settings.VIPONLY and not game.Workspace.vipServer.Value then
		warn("VIPONLY setting is enabled, and this is a Public server. Pausing..")
		repeat wait(1) until not game.Workspace.vipServer.Value
	end
end

function DQAL.CreateLobby()
	DQAL.Prnt("Creating a "..Settings.LobbyType.." Lobby..")
	if Settings.LobbyType == "Dungeon" then
		SRemote("createLobby", Settings.MapName, Settings.Difficulty, 0, Settings.Hardcore, Settings.Private, Settings.WaveDefense)
		LobbyFolder = game.Workspace.games.inLobby:WaitForChild(MyName, 30)
	elseif Settings.LobbyType == "BossRaid" then
		SRemote("createBossLobby", Settings.BRTier, Settings.Private, Settings.BRTierReq)
		LobbyFolder = game.Workspace.bossLobbies:WaitForChild(MyName, 30)
	end
	DQAL.Prnt(Settings.LobbyType.." Lobby created.")
end

function DQAL.JoinLobby()
	DQAL.Prnt("Looking for "..Settings.LobbyHost.."'s Lobby..")
	if Settings.LobbyType == "Dungeon" then
		LobbyFolder = game.Workspace.games.inLobby:WaitForChild(Settings.LobbyHost, 1200)
	elseif Settings.LobbyType == "BossRaid" then
		LobbyFolder = game.Workspace.bossLobbies:WaitForChild(Settings.LobbyHost, 1200)
	end
	DQAL.Prnt(Settings.LobbyHost.."'s "..Settings.LobbyType.." Lobby found.")
	DQAL.Prnt("Joining "..Settings.LobbyHost.."'s Lobby..")

	if Settings.LobbyType == "Dungeon" then
		SRemote("joinDungeon", Settings.LobbyHost)
		game.Workspace.games.inLobby[Settings.LobbyHost]:WaitForChild(MyName, 30)
	elseif Settings.LobbyType == "BossRaid" then
		local raidHost = game.Workspace.bossLobbies[Settings.LobbyHost]
		SRemote("playerJoinBossLobby", raidHost)
		game.Workspace.bossLobbies[Settings.LobbyHost]:WaitForChild(MyName, 30)
	end

	DQAL.Prnt("Lobby joined. Waiting for "..Settings.LobbyHost.." to start game..")
	while game.PlaceId == 2414851778 do	wait(.5) end
end

function DQAL.Whitelist()
	if #Settings.WLParty <= 0 then return end
	local lobbyPlyrs, plyrsAdded, plyrsOnline, partyReady = {},{},{},false
	DQAL.Prnt("Waiting to whitelist ["..#Settings.WLParty.."] players..")
	repeat
		wait()
		plyrsOnline = game:GetService("Players"):GetPlayers()
		for _,p in pairs(plyrsOnline) do
			if table.find(Settings.WLParty, p.Name) and not table.find(lobbyPlyrs, p.Name) and not table.find(plyrsAdded, p.Name) then
				if Settings.LobbyType == "Dungeon" then
					SRemote("addPlayerToWhitelist", p.Name)
				elseif Settings.LobbyType == "BossRaid" then
					SRemote("addPlayerToBossWhitelist", p.Name)
				end
				table.insert(plyrsAdded, p.Name)
				DQAL.Prnt(p.Name.." whitelisted. ["..#plyrsAdded.."/"..#Settings.WLParty.."]")
			end
		end
		lobbyPlyrs = LobbyFolder:GetChildren()
		if Settings.LobbyType == "Dungeon" then
			if #Settings.WLParty == #lobbyPlyrs-2 then partyReady = true end
		elseif Settings.LobbyType == "BossRaid" then
			if #Settings.WLParty == #lobbyPlyrs-1 then partyReady = true end
		end
	until(partyReady and (#Settings.WLParty == #plyrsAdded))
	DQAL.Prnt("All players whitelisted and ready.")
end

function DQAL.StartDungeon()
	DQAL.Prnt("Starting game..")
	if Settings.LobbyType == "Dungeon" then
		SRemote("startDungeon")
	elseif Settings.LobbyType == "BossRaid" then
		SRemote("startBossRaid")
	end
	DQAL.Prnt("Waiting to teleport out..")
	while game.PlaceId == 2414851778 do	wait(.5) end
end

if Settings.Enabled then
	DQAL.PreRun()
	if Settings.LobbyHost == MyName then
		DQAL.CreateLobby()
		DQAL.Whitelist()
		DQAL.StartDungeon()
	else
		DQAL.JoinLobby()
	end
end

-- // End of Dungeon Quest AutoLobby // --


AutolobbyDQ-AFCheck.lua
--[[
Dungeon Quest Autolobby->Autofarm Run Check
  This file simply prevents the Autofarm script from running by ensuring it doesn't continue to it by checking the player's UserId.
  For example, if you are using the same PC to run 2 Roblox instances and host on one account and join it on the other. You would put the
  User Id of the account getting carried into this file so that it only runs the Autolobby code, and not the Autofarm code.
  You can get your User Id by visiting your Profile link on the Roblox account and then copying the number in the internet url.
  Example: https://www.roblox.com/users/USERIDNUMBER/profile
]]

--/ Replace the number here with the Account's ID number /--
local useridnum = 000000000
--/ Do not edit below this line /--

while not game.Players.LocalPlayer do wait() end
if game.Players.LocalPlayer.UserId == useridnum then
	print("AFCheck preventing any further script executions..")
	while game.Players.LocalPlayer.UserId == useridnum do
		wait()
	end
end


AutolobbyDQ.lua
-----------------------------------------------------------------------------
-- [[ Dungeon Quest AutoLobby Addon v1.3 - by dns (dns#3420 on discord) ]] --
--               https://github.com/dns2/RBX-DQ-AutoLobby                  --
-----------------------------------------------------------------------------
--[[ 	FOR: Adds Carrying (Auto-hosting and/or Auto-joining) functionality to Blake's Autofarm.
		TO USE:
			Configure the settings below, put autolobby file in exploit's autoexec folder, put autofarm file in exploit's 'workspace' folder.
		SUPPORT: If you need help with using/configuring this, then message me (dns#3420) on discord, or use link below.
			* Full Configuration and Usage info is available at https://github.com/dns2/RBX-DQ-AutoLobby/blob/master/README.md
		NOTES:
			This will not teleport you to VIP servers. The VIPONLY option is only for usage with the 'AutoRestart' program.
			Usernames, Map name, & Difficulty are all case-sensitive. *Example: "player1" is not the same as "Player1")		]]

-- [ CONFIG ]
Settings = {
	--> [ Script ]
		Enabled = true,						-- false = disable/skip Autolobby
		VIPONLY = false,					-- true = Suspend everything unless on a VIP Server
	--> [ Hosting ]
		LobbyHost = "HostName",				-- Username of player that will be creating/hosting the game lobby. Will also accept "" to be host
		WLParty  = {"Player1"},				-- Username(s) to whitelist. Example: {"Player1","Player2","Player3"} | Will also accept {} for solo
		LobbyType = "Dungeon",				-- "Dungeon" or "BossRaid"
	--> [ Dungeon ]
		MapName = "Orbital Outpost",
		Difficulty = "Nightmare",
		Hardcore = true,
		WaveDefense = false,				-- Keep this false. (Wave defence doesn't currently work on Blake's Autofarm)
	--> [ BossRaid ]
		BRTier = 30,						-- I'll add an Auto feature for this in a future update
		BRTierReq = 0,						-- Leave at 0. Reserved for future update
	--> [ Misc ]
		Private = true,						-- Keep this true. Reserved for future update
		DisableMusic = true,				-- true = Music disabled | false = Music enabled
		DisableTrade = true,				-- true = Trading disabled | false = Trading enabled
		DebugOutput = true,					-- Enable or Disable script output to Roblox Developer Console
		NameOfAutofarmFile = "Blakes_P_DQ_Autofarm_V5.1.lua",	-- EXACT name of the autofarm lua file
		RunAfOnlyAsHost = true				-- true = Will only run Autofarm when you're the host | false = Will always run Autofarm when Dungeon loads
	} -- [ END CONFIG ] --> (Do not edit below this line)

	DQAL = {}
	SRemote = {
		rFile = nil,
		invoke = {"createLobby","createBossLobby","joinDungeon","playerJoinBossLobby"},
		fire = {"loadPlayerCharacter","changeMute","changeDnD","addPlayerToWhitelist","addPlayerToBossWhitelist","startDungeon","startBossRaid"}}
	setmetatable(SRemote, {__call =
	function(tbl, rScript, ...)
		if not tbl.rFile then tbl.rFile = game:GetService("ReplicatedStorage"):FindFirstChild("remotes") end
		if table.find(tbl.invoke, rScript) then tbl.rFile:FindFirstChild(rScript):InvokeServer(...) return end
		if table.find(tbl.fire, rScript) then tbl.rFile:FindFirstChild(rScript):FireServer(...) return end
	end})

	function DQAL.Prnt(...)
		if Settings.DebugOutput then print("[DQAL]:",...) end
	end

	function DQAL.Die(p, ...)
		if p then DQAL.Prnt(...) end
		Settings = nil
		DQAL = nil
		script:Destroy()
		script.Disabled = true
		error("'I refused to die, but alas, my creator willed it so. Therefore, it shall be done.' [Coderonians:187]", 0)
	end

	function DQAL.PreRun()
		DQAL.Prnt("Dungeon Quest AutoLobby initialized. Waiting for game..")
		repeat wait() until game:IsLoaded()
		local place = game.PlaceId
		if place ~= 2414851778 then
			if not DQAL.CheckPlaceID(place) then
				DQAL.Die(true, "Not a known Dungeon Quest Lobby/Dungeon. Exiting..")
			elseif DQAL.CheckPlaceID(place) then
				return 2 -- Determine whether to load Autofarm
			end
		end
		DQAL.Prnt("Game is loaded. Waiting for character..")
		SRemote("loadPlayerCharacter", true)
		MyChar = game.Players.LocalPlayer.Character or game.Players.LocalPlayer.CharacterAdded:wait()
		DQAL.Prnt("Character loaded. Continuing..")
		MyName = game.Players.LocalPlayer.Name
		if Settings.LobbyHost == "" then Settings.LobbyHost = MyName end
		if Settings.DisableMusic then SRemote("changeMute") end
		if Settings.DisableTrade then SRemote("changeDnD") end
		if Settings.VIPONLY and not game.Workspace.vipServer.Value then
			warn("VIPONLY setting is enabled, and this is a Public server. Pausing..")
			repeat wait(1) until not game.Workspace.vipServer.Value
		end
		return 1 -- Allow Autolobby to continue
	end

	function DQAL.CreateLobby()
		DQAL.Prnt("Creating a "..Settings.LobbyType.." Lobby..")
		if Settings.LobbyType == "Dungeon" then
			SRemote("createLobby", Settings.MapName, Settings.Difficulty, 0, Settings.Hardcore, Settings.Private, Settings.WaveDefense)
			LobbyFolder = game.Workspace.games.inLobby:WaitForChild(MyName, 60)
		elseif Settings.LobbyType == "BossRaid" then
			SRemote("createBossLobby", Settings.BRTier, Settings.Private, Settings.BRTierReq)
			LobbyFolder = game.Workspace.bossLobbies:WaitForChild(MyName, 60)
		end
		DQAL.Prnt(Settings.LobbyType.." Lobby created.")
	end

	function DQAL.JoinLobby()
		DQAL.Prnt("Looking for "..Settings.LobbyHost.."'s Lobby..")
		if Settings.LobbyType == "Dungeon" then
			LobbyFolder = game.Workspace.games.inLobby:WaitForChild(Settings.LobbyHost, 1200)
		elseif Settings.LobbyType == "BossRaid" then
			LobbyFolder = game.Workspace.bossLobbies:WaitForChild(Settings.LobbyHost, 1200)
		end
		DQAL.Prnt(Settings.LobbyHost.."'s "..Settings.LobbyType.." Lobby found.")
		wait()
			function sendJoin()
				DQAL.Prnt("Joining "..Settings.LobbyHost.."'s Lobby..")
				if Settings.LobbyType == "Dungeon" then
					SRemote("joinDungeon", Settings.LobbyHost)
					game.Workspace.games.inLobby[Settings.LobbyHost]:WaitForChild(MyName, 10)
				elseif Settings.LobbyType == "BossRaid" then
					local raidHost = game.Workspace.bossLobbies[Settings.LobbyHost]
					SRemote("playerJoinBossLobby", raidHost)
					game.Workspace.bossLobbies[Settings.LobbyHost]:WaitForChild(MyName, 10)
				end
			end
			function confirmJoin()
				if Settings.LobbyType == "Dungeon" then
					if not game.Workspace.games.inLobby[Settings.LobbyHost]:FindFirstChild(MyName) then sendJoin() end
				elseif Settings.LobbyType == "BossRaid" then
					if not game.Workspace.bossLobbies[Settings.LobbyHost]:FindFirstChild(MyName) then sendJoin() end
				end
			end
		sendJoin()
		DQAL.Prnt("Lobby joined. Waiting for "..Settings.LobbyHost.." to start game..")
		while game.PlaceId == 2414851778 do
			wait(5)
			confirmJoin()
		end
	end

	function DQAL.Whitelist()
		if #Settings.WLParty <= 0 then return end
		local lobbyPlyrs, plyrsAdded, plyrsOnline, partyReady = {},{},{},false
		DQAL.Prnt("Waiting to whitelist ["..#Settings.WLParty.."] players..")
		repeat
			wait()
			plyrsOnline = game:GetService("Players"):GetPlayers()
			for _,p in pairs(plyrsOnline) do
				if table.find(Settings.WLParty, p.Name) and not table.find(lobbyPlyrs, p.Name) and not table.find(plyrsAdded, p.Name) then
					if Settings.LobbyType == "Dungeon" then
						SRemote("addPlayerToWhitelist", p.Name)
					elseif Settings.LobbyType == "BossRaid" then
						SRemote("addPlayerToBossWhitelist", p.Name)
					end
					table.insert(plyrsAdded, p.Name)
					DQAL.Prnt(p.Name.." whitelisted. ["..#plyrsAdded.."/"..#Settings.WLParty.."]")
				end
			end
			lobbyPlyrs = LobbyFolder:GetChildren()
			if Settings.LobbyType == "Dungeon" then
				if #Settings.WLParty == #lobbyPlyrs-2 then partyReady = true end
			elseif Settings.LobbyType == "BossRaid" then
				if #Settings.WLParty == #lobbyPlyrs-1 then partyReady = true end
			end
		until(partyReady and (#Settings.WLParty == #plyrsAdded))
		DQAL.Prnt("All players whitelisted and ready.")
	end

	function DQAL.StartDungeon()
		DQAL.Prnt("Starting game..")
		if Settings.LobbyType == "Dungeon" then
			SRemote("startDungeon")
		elseif Settings.LobbyType == "BossRaid" then
			SRemote("startBossRaid")
		end
		DQAL.Prnt("Waiting to teleport out..")
		while game.PlaceId == 2414851778 do	wait(.5) end
	end

	function DQAL.CheckPlaceID(_id)
		local Dungeons = {
			2606294912, -- Desert Temple
			2743806150, -- Winter Outpost
			2988891534, -- Pirate Island
			3041739550, -- King's Castle
			3119903031, -- The Underworld
			3277965370, -- Samurai Palace
			3488584454,	-- The Canals
			3737465474, -- Ghastly Harbor
			4113459044, -- Steampunk Sewers
			4628698373, -- Orbital Outpost
			4865331948, -- Egg Island
		}
		if table.find(Dungeons, _id) then
			return true
		else
			return false
		end
	end

	if Settings.Enabled then
		local task = DQAL.PreRun()
		if task == 1 then
			if Settings.LobbyHost == MyName then
				DQAL.CreateLobby()
				DQAL.Whitelist()
				DQAL.StartDungeon()
			elseif Settings.LobbyHost ~= MyName then
				DQAL.JoinLobby()
			end
		elseif task == 2 then
			MyName = game.Players.LocalPlayer.Name
			if Settings.LobbyHost == MyName then
				DQAL.Prnt("Joined Dungeon as Host! Loading '"..Settings.NameOfAutofarmFile.."'..")
				do loadfile(Settings.NameOfAutofarmFile)() end
			elseif Settings.LobbyHost ~= MyName then
				if not Settings.RunAfOnlyAsHost then
					DQAL.Prnt("Joined Dungeon as Alt/Carry! Loading '"..Settings.NameOfAutofarmFile.."'..")
					do loadfile(Settings.NameOfAutofarmFile)() end
				elseif Settings.RunAfOnlyAsHost then
					DQAL.Prnt("Joined Dungeon as Alt/Carry!")
					return
				end
			end
		end
	end

	-- // End of Dungeon Quest AutoLobby // --
